modules/gdnative/include/gdnative/variant.h

```diff
@@ -37,12 +37,11 @@ extern "C" {
 
 #include <stdint.h>
 
-#define GODOT_VARIANT_SIZE (16 + sizeof(void *))
-
 #ifndef GODOT_CORE_API_GODOT_VARIANT_TYPE_DEFINED
 #define GODOT_CORE_API_GODOT_VARIANT_TYPE_DEFINED
 typedef struct {
-	uint8_t _dont_touch_that[GODOT_VARIANT_SIZE];
+	int64_t _dont_touch_that[2]; // uses workaround to correct calculate sizeof(Variant) for armv7 with memory aligment to 8 bytes
```

Comment:
I think all core types should be aligned to 8 bytes via attributes. Doing that is a bit different on all compilers, but I think it's clearer than this.

Reply:
Do you mean updating core types in Godot itself? Variant actually already had this attribute, but only for GCC. Do you suggest ensuring that Variant is 24 bytes on all platforms instead of this fix?

Reply:
No, I meant that at the moment non of the GDNative wrapper types have alignment specified and are represented as byte arrays, so their alignment is always `1`.

On x64 this is not much of a problem, but unaligned reads can be fatal on other platforms. Instead of converting the size to `int64_t[2] + int` I would keep the byte array but add explicit alignment.

The size still needs to be determined properly, of course, maybe the most clear way is actually to have just a bunch of `#if PLATFORM_X` to specify the size directly.

Reply:
The alignment is `1` only if you specify that the struct is packed (which is another attribute). Otherwise compilers make them aligned by default, depending on the target platform. I'm still not sure why we should mess with alignment in these wrapper types. Their sole purpose is to have size equal to the types from Godot. So as long as it is fulfilled, I don't think we should add more complexity to their definitions.

Static checks added in this commit ensure that the sizes are equal on any platform Godot is compiled for.

Reply:
As far as I know, the alignment is always the biggest alignment of all the members. In this case all members are just `uint8_t`, so the alignment is always `1`. There is no way the compiler knows that the C wrapper types need to fit data that contains pointers or multi-byte integers.

https://godbolt.org/z/q7M4Gg

Additionally to size checks that you added we should also add alignment checks. In the GDNative C wrappers pointers to the C types are often just casted to pointers of the C++ type. Unaligned access can be a problem, even if the size matches.

Reply:
The change makes the largest member int64_t which ensures 8-byte alignment. I don't understand why you insist on adding platform- or compiler- specific ifdefs, it would look more messy to me.

Alignment checks isn't a static thing. If allocator provides aligned addresses and the data itself is aligned (size is divisble by the required alignment), you'll have aligned pointers.

Reply:
Ah, my apologies, I didn't check the snippet you provided, and still not used to the fact that C++11 is a thing in Godot now. My question is: since, as you said, pointers to C wrappers are actually pointers to previously constructed C++ types, how can they be unaligned? They are ever allocated directly, or are they?

Reply:
My apologies, I've now realized what you meant by alignment checks and agree they should be added. Still not sure about the ifdefs.

Reply:
I have no strong feelings about the `ifdef`s, I just think a solution that is clear and works on all architectures would be good.

The `sizeof(void *) + N` or hardcoded values can work, but it's less clear to read for me personally than a list of exceptions for example.

But I think we can deal with the size somehow. The mismatch of alignment however is something that keeps me awake at night :sweat_smile:

Reply:
Well, size is already dealt with - the CI build passes for all platforms, which means the static checks have passed there. We should add alignment checks and see the result. I think it can be tweaked by simply using specifically-sized types (`int32_t`, `int64_t`) instead of byte arrays.